### DISCUSSION TITLE
Rounding error between attribute price and shopping cart price

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#5829] [Ideas]
First reported here in 2015
https://www.zen-cart.com/showthread.php?216906-Rounding-error-between-attribute-price-and-shopping-cart-price

Bug example: Attribute price is 10 on Product Info page, but when added to cart it is calculated at 9.99.
This is evident to users of Dynamic Price Updater where it updates with the wrong price on the page.

Steps to duplicate this

1) Currency Euro (probably not relevant).
2) Tax zone all countries, all zones
3) Tax rate 21%
4) Configuration->My Store: show prices with tax.
5) Edit the product attributes of:
Test 10% by Attrib
Hewlett Packard - by attributes SALE (pid 36).
Set white to 8.2644 (10 euros -21% tax).
6) View in shop: white attribute = 10 euros
7) Add to cart = 9.99.

**Comment by drbyte** [comment]
I've pushed #4140 as a starting point to explore this issue. 

However, I've not tested it with fractional tax rates, nor fractional quantities for products or attributes. I'll leave those tests to you since you have real-world cases to explore those.  Please post the examples you test for.

**Comment by torvista** [comment]
Initial test
tax 21%
show prices with tax
product base price 0, priced by attributes

Pre fix:
Shown attribute prices were 41, 61, 71 
DPU/Cart prices were 40.99, 61, 71

Post fix: 41, 61, 71. 

So it seems to work!

I have no fractional aspects, so I'll get this online asap.

Many thanks for reviewing this so quickly. Sort of...:satisfied:

**Comment by mc12345678** [comment]
Oh no, here we go with the possible penny rounding error again...
@lat9, care to weigh on any of this?

**Comment by mc12345678** [comment]
Tried this on ZC 1.5.7, small problem if order 1000 of an item, then the shopping cart product price comes out correct (10000.00), but the sub-total comes out incorrect: 9,999.92.

Of course I wonder why the price was entered as 8.2644 instead of 8.2645 which is the result of calculating the non-taxed price of a product that is to have a final price of 10 with a tax rate of 21%. This gives an initial price of 8.264462809917355 which would have been 8.2645 if keeping it to 4 significant values.

Although if I used 8.2645 and the 1000 quantity of product then the sub-total ends up being a little off in a different way: 10000.05..

This whole situation though is something that I've tried to have a discussion about... How when the currency only has 2 significant figures can the product's price honestly be something other than a value that has the same number of significant figures.. E.g., if someone that doesn't have to pay VAT or the publicized price but instead would pay the "calculated" price. How does one collect 8.2644 from the customer? And if instead 8.26 was collected, what happens if they then do this multiple times, say 10 times of buying 1 at a time at 8.26 without having to pay the VAT? Buy that product 100 times over in that way and then if the VAT were needed to be calculated off of those product: 8.26*100 = 826. Then 826 * 1.21 = 999.46...  Like I'm wondering if the fix isn't either something else with the software or something else with the data entry or both... Perhaps something like a price entered in the admin is to be that of the final price or of the pre-tax price.  In this way, if a store entered 10.00 it would represent the final price, but if they entered 8.2644 or 8.2645 it represented the untaxed price and that later prices would be calculated based off of that information.

Or maybe I just always try to find ways that this "math" stuff may not work...

**Comment by torvista** [comment]
> How when the currency only has 2 significant figures can the product's price honestly be something other than a value that has the same number of significant figures

I see the logic in this restriction.
Prevents having nicely rounded final prices in some cases, but it's evident that this long-running saga of chasing rounding errors in multiple places is a moving target, depending on store configuration / not a result of real bugs as such.
Is it solvable without putting such a simple restriction in place?
How do other shops deal with this?

**Comment by torvista** [comment]
...I've concluded that since customers (whether B2B inside the EU, or export) can buy ex-vat, this is the actual "real price" of a product and so logically this should be stored in (or entered in/extracted from) the database in the format concurrent with the currency in use, in terms of the decimal places.

The current method of being able to enter a +vat price on the product page and allowing the -vat price to be calculated and stored with 4 decimal places goes against this, and seems wrong in this context.

I guess this may be why I see things in other shops priced at what I consider odd prices such as 123.42 instead of being rounded in some aspec:t: the +vat price is calculated from a base price with two decimals.

Perhaps enforcing this method/policy of storing the base price would cure the rounding errors?

**Comment by torvista** [comment]
The only subsequent rounding error I have seen is this.
Product with price set at 159->stored at 131.405. VAT 21%.
Cart shows 159. 
Payment page subtotal shows 159.01.
Generated from this
```
$this->output[] = array('title' => $this->title . ':',
                              'text' => $currencies->format($order->info['subtotal'], true, $order->info['currency'], $order->info['currency_value']),
                              'value' => $order->info['subtotal']);
```
Which is this:

> [0] => Array
>         (
>             [title] => Subtotal:
>             [text] => 159,01€
>             [value] => 159.0061
>         )

Reinforces my view in the previous post: I think it is impossible to have no rounding errors is all possible cases without limiting the base price decimals as per the currency in use.

**Comment by lat9** [comment]
> > How when the currency only has 2 significant figures can the product's price honestly be something other than a value that has the same number of significant figures
> 
> I see the logic in this restriction. Prevents having nicely rounded final prices in some cases, but it's evident that this long-running saga of chasing rounding errors in multiple places is a moving target, depending on store configuration / not a result of real bugs as such. Is it solvable without putting such a simple restriction in place? How do other shops deal with this?
Sorry, but I just saw this one.  Consider a site that sells bulk nails. with a minimum purchase of 1000 where each nail is priced at $0.0099.  I'm sure that that would wreck havoc on the computations, especially if the associated tax's decimals was also in the thousandths (e.g. 5.125%).

**Comment by torvista** [comment]
Recently I did a mod to do trade pricing, first time I have crossed paths with the calculation of prices, and it seemed a scary mess of duplication moving through the shopfront and cart processing. It seemed to me to need refactoring to simplify things before trying to solve problems like this, way out of my pay grade.
